### PR TITLE
Rework checks and fix signed overflow check

### PR DESCRIPTION
**Pull request by sallaigy**
This PR contains some refactoring done to the check instrumentation infrastructure (splitting `DefaultChecks.cpp` into several files, one for each check) and fixing some issues with the highly experimental signed overflow check we've had beforehand. During implementation some bugs surfaced up, this PR also includes a fix for those:

1. Invalid CFA was built if control flow jumped two levels out of its current loop, this should be fixed by the newly introduced `LoopExitCanonizationPass`, which eliminates such constructs.
2. Too large bit-vector literals crashed the Z3 solver backend.

### CODE REVIEW

**Review by radl97 — COMMENTED** [review]
I haven't touched most of this code, so I do not understand most of the changes here.
I only have some questions, but they may not be so important :) (It seems that the PR contains more than described)

File: src/LLVM/ClangFrontend.cpp [hunk 1]
```diff
@@ -240,6 +240,9 @@ void ClangOptions::addSanitizerFlag(llvm::StringRef flag)
 void ClangOptions::createArgumentList(std::vector<std::string>& args)
 {
     if (!mSanitizerFlags.empty()) {
-        args.emplace_back("-fsanitize-trap=" + llvm::join(mSanitizerFlags, ","));
+        auto sanitizerNames = llvm::join(mSanitizerFlags, ",");
+
+        args.emplace_back("-fsanitize=" + sanitizerNames);
+        args.emplace_back("-fno-sanitize-recover=" + sanitizerNames);
```
[the remark this answers is not in the document]
Comment (sallaigy) [reply src/LLVM/ClangFrontend.cpp:246]: `-fsanitize-trap` inserts a function call to `llvm.trap`, whereas `-fno-sanitize-recover` will insert a call to a function more specific to the error, e.g. `__ubsan_handle_add_overflow_abort`, so it's easier to disambiguate between different error calls if we write another check that uses UBSan.

[the remark this answers is not in the document]
Comment (sallaigy) [reply src/LLVM/ClangFrontend.cpp:246]: `-fsanitize-trap` inserts a call to a generic `llvm.trap` function, whereas with `-fno-sanitize-recover` we get a call to a more specific abort function, e.g. `__ubsan_handle_add_overflow_abort`, so it is easier to disambiguate between such calls (especially if we ever write another check that uses UBSan instrumentation).

File: src/LLVM/Instrumentation/Checks/SignedIntegerOverflowCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    if (callee == nullptr) {
+        return false;
+    }
+
+    auto name = callee->getName();
+
+    return name == "__ubsan_handle_add_overflow_abort"
+        || name == "__ubsan_handle_sub_overflow_abort"
+        || name == "__ubsan_handle_mul_overflow_abort"
+        || name == "__ubsan_handle_divrem_overflow_abort";
```
[the remark this answers is not in the document]
Comment (sallaigy) [reply src/LLVM/Instrumentation/Checks/SignedIntegerOverflowCheck.cpp:118]: Indeed, I will remove these.

File: include/gazer/LLVM/Instrumentation/Intrinsics.h [hunk 3]
```diff
@@ -50,7 +50,7 @@ class GazerIntrinsic
 
     enum class Overflow
     {
-        SAdd, UAdd, SSub, USub, SMul, UMul, SDiv, Shl
```
[the remark this answers is not in the document]
Comment (sallaigy) [reply include/gazer/LLVM/Instrumentation/Intrinsics.h:53]: Deciding if these will overflow seemed a little bit tricikier and wanted to handle them later. However, the handler code will probably crash if we have a div overflow intrinsic, so I will make the update.

**Pull request merged**